src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs

```diff
@@ -47,88 +47,70 @@ public virtual NextVersion FindVersion()
-        nextVersion.BaseVersion.SemanticVersion.BuildMetaData = this.mainlineVersionCalculator.CreateVersionBuildMetaData(nextVersion.BaseVersion.BaseVersionSource);
+        var baseVersion = nextVersion.BaseVersion;
+        var preReleaseTagName = nextVersion.Configuration.GetBranchSpecificTag(this.log, Context.CurrentBranch.Name.Friendly, baseVersion.BranchNameOverride);
+
         SemanticVersion semver;
         if (Context.Configuration.VersioningMode == VersioningMode.Mainline)
         {
-            semver = this.mainlineVersionCalculator.FindMainlineModeVersion(nextVersion.BaseVersion);
+            semver = this.mainlineVersionCalculator.FindMainlineModeVersion(baseVersion);
         }
         else
         {
-            if (taggedSemanticVersion?.BuildMetaData == null || (taggedSemanticVersion.BuildMetaData?.Sha != nextVersion.BaseVersion.SemanticVersion.BuildMetaData.Sha))
+            var baseVersionBuildMetaData = this.mainlineVersionCalculator.CreateVersionBuildMetaData(baseVersion.BaseVersionSource);
+
+            if (baseVersionBuildMetaData == null || baseVersionBuildMetaData.Sha != nextVersion.IncrementedVersion.BuildMetaData?.Sha)
             {
                 semver = nextVersion.IncrementedVersion;
-                semver.BuildMetaData = this.mainlineVersionCalculator.CreateVersionBuildMetaData(nextVersion.BaseVersion.BaseVersionSource);
             }
             else
             {
-                semver = nextVersion.BaseVersion.SemanticVersion;
+                semver = baseVersion.SemanticVersion;
+            }
+
+            semver.BuildMetaData = baseVersionBuildMetaData;
+
+            var lastPrefixedSemver = this.repositoryStore
+                .GetVersionTagsOnBranch(Context.CurrentBranch, Context.Configuration.TagPrefix)
+                .Where(v => MajorMinorPatchEqual(v, semver) && v.PreReleaseTag?.HasTag() == true)
+                .FirstOrDefault(v => v.PreReleaseTag?.Name?.IsEquivalentTo(preReleaseTagName) == true);
+
+            if (lastPrefixedSemver != null)
+            {
+                semver.PreReleaseTag = lastPrefixedSemver.PreReleaseTag;
             }
         }
 
-        var hasPreReleaseTag = semver.HasPreReleaseTagWithLabel;
-        var tag = nextVersion.Configuration.Tag;
-        var branchConfigHasPreReleaseTagConfigured = !tag.IsNullOrEmpty();
-        var preReleaseTagDoesNotMatchConfiguration = hasPreReleaseTag && branchConfigHasPreReleaseTagConfigured && semver.PreReleaseTag?.Name != tag;
-        if (semver.PreReleaseTag?.HasTag() != true && branchConfigHasPreReleaseTagConfigured || preReleaseTagDoesNotMatchConfiguration)
+        if (Context.CurrentCommitTaggedVersion != null && semver.CompareTo(Context.CurrentCommitTaggedVersion) == 0)
         {
-            UpdatePreReleaseTag(new(nextVersion.Branch, nextVersion.Configuration), semver, nextVersion.BaseVersion.BranchNameOverride);
+            if (semver.BuildMetaData != null)
```

Comment:
This is always true because you have set it before with a not null value (see comment before). Did we lose a use case!? It's fine with me if you delete this check as far the unit tests are green. ;)

Reply:
I would say: Change the data type of the property PreReleaseTag and BuildMetaData to a non-nullable type in the class SemanticVersion. ;)

Reply:
Done 🤠